xkeys.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2022 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nkeys
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/binary"
+	"io"
+
+	"golang.org/x/crypto/curve25519"
+	"golang.org/x/crypto/nacl/box"
+)
+
+// This opackage will support safe use of X25519 keys for asymmetric encryption.
+// We will be compatible with nacl.Box, but generate random nonces automatically.
+// We may add more advanced options in the future for group recipients and better
+// end to end algorithms.
+
+// CurveKeyPair provides the central interface to xkeys used for encryption.
+type CurveKeyPair interface {
+	Seed() ([]byte, error)
+	PublicKey() (string, error)
+	PrivateKey() ([]byte, error)
+	Seal(input []byte, recipient string) ([]byte, error)
+	SealWithRand(input []byte, recipient string, rr io.Reader) ([]byte, error)
+	Open(input []byte, sender string) ([]byte, error)
+	Wipe()
+}
+
+const (
+	curveKeyLen    = 32
+	curveDecodeLen = 35
+	curveNonceLen  = 24
+)
+
+type ckp struct {
+	seed [curveKeyLen]byte // Private raw key.
+}
+
+// CreateUser will create a User typed KeyPair.
+func CreateCurveKeys() (CurveKeyPair, error) {
+	return CreateCurveKeysWithRand(rand.Reader)
+}
+
+// CreateUser will create a User typed KeyPair with specified rand source.
+func CreateCurveKeysWithRand(rr io.Reader) (CurveKeyPair, error) {
+	var kp ckp
+	_, err := io.ReadFull(rr, kp.seed[:])
+	if err != nil {
+		return nil, err
+	}
+	return &kp, nil
+}
+
+// Will create a curve key pair from seed.
+func FromCurveSeed(seed []byte) (CurveKeyPair, error) {
+	pb, raw, err := DecodeSeed(seed)
+	if err != nil {
+		return nil, err
+	}
+	if pb != PrefixByteCurve || len(raw) != curveKeyLen {
+		return nil, ErrInvalidCurveSeed
+	}
+	var kp ckp
+	copy(kp.seed[:], raw)
+	return &kp, nil
+}
+
+// Seed will return the encoded seed.
+func (pair *ckp) Seed() ([]byte, error) {
+	return EncodeSeed(PrefixByteCurve, pair.seed[:])
+}
+
+// PublicKey will return the encoded public key.
+func (pair *ckp) PublicKey() (string, error) {
+	var pub [curveKeyLen]byte
+	curve25519.ScalarBaseMult(&pub, &pair.seed)
+	key, err := Encode(PrefixByteCurve, pub[:])
+	return string(key), err
+}
+
+// PrivateKey will return the encoded private key.
+func (pair *ckp) PrivateKey() ([]byte, error) {
+	return Encode(PrefixBytePrivate, pair.seed[:])
+}
```

Comment:
**BUG: CRYPTOGRAPHIC DANGER**

This is wrong.  The seed is not the private key.  See how the existing code uses `ed25519.GenerateKey()` on the seed: the seed is effectively the random source seed for the key generator.

Reply:
In X25519 I believe it is.

https://cs.opensource.google/go/x/crypto/+/refs/tags/v0.3.0:nacl/box/box.go;l=61

Reply:
https://go.dev/play/p/JFCdtu-pMa4